### ISSUE TITLE
SIGSEGV caused by Excerpts safeLimit being twice (mappedSize - safeCapacity) in MappedFile

### ISSUE THREAD

**Issue by gity** [OpenHFT/Chronicle-Queue#535]
When writing to `SingleChroncileQueueBuilder.binary()` bytes of 16777216 using `appender.writeBytes(b -> b.write(bytes16M))` the 5th write can cause a SIGSEGV because the write seems to be writing past the mapped size.

This seems to be becuase the 
[safeLength](https://github.com/OpenHFT/Chronicle-Queue/blob/705a8d3d47026f25754ea14db44c05c3a2cdb2cc/src/main/java/net/openhft/chronicle/queue/impl/single/SingleChronicleQueueExcerpts.java#L364) in `SingleChronicleQueueExcerpts` is `queue.overlapSize()` while [mappedSize](https://github.com/OpenHFT/Chronicle-Bytes/blob/master/src/main/java/net/openhft/chronicle/bytes/MappedFile.java#L342) in Chronicle-Bytes `MappedFile` is `chunkSize + overlapSize` and [safeCapacity](https://github.com/OpenHFT/Chronicle-Bytes/blob/168ebbca0d1096163e244e0ef110967467db0ed2/src/main/java/net/openhft/chronicle/bytes/MappedFile.java#L346) is `this.chunkSize + overlapSize / 2`.

It seems that preferably that `safeCapacity = this.chunkSize` in `MappedFile` or otherwise `safeLength = queue.overlapSize() / 2` in `SingleChronicleQueueExcerpts`

**Comment by RobAustin** [comment]
Please provide a failing test, and re-test against the latest, As  I attempted to reproduce your issue but the following test passes 

````
    @Test
    public void testMappedFileSafeLimit() throws Exception {
        Path tempFile = Files.createTempFile("x", "y");
        MappedBytes b = MappedBytes.mappedBytes(tempFile.toFile(), 1 << 16);
        Wire wire = WireType.BINARY.apply(b);

        try {
            for (int i = 0; i < 5; i++) {
                try (DocumentContext dc = wire.writingDocument()) {
                    Bytes<?> bytes = dc.wire().bytes();
                    bytes.writeSkip((16 << 20) - 1);
                    bytes.writeByte((byte) 1);
                }

            }
        } finally {
            b.release();
        }
    }

````

**Comment by RobAustin** [comment]
@gity I've closed this issue for now, but if you find there is still an issue then please resubmit with a failing test

**Comment by gity** [comment]
It's still an issue. The problem is there interplay between Chronicle and MappedFile.

Try creating a Chronicle Binary Queue with a block size of 65536. Call writeBytes 5 times with an array of 40,000 bytes each time. Close the queue and reopen and try and read the bytes you wrote.

**Comment by RobAustin** [comment]
I suggest you increase your block size to x8 of your array size

```
package net.openhft.chronicle.queue;

import net.openhft.chronicle.queue.impl.single.SingleChronicleQueueBuilder;
import net.openhft.chronicle.wire.DocumentContext;
import net.openhft.chronicle.wire.WireType;
import org.junit.Assert;

import java.io.File;
import java.util.Arrays;

/**
 * Created by Rob Austin
 */
public class Test extends ChronicleQueueTestBase {

    private static final int ARRAY_SIZE = 40_000;
    private static final int BLOCK_SIZE = ARRAY_SIZE * 8;

    @org.junit.Test
    public void testMappedFileSafeLimit() throws Exception {
        byte[] data = new byte[ARRAY_SIZE];
        Arrays.fill(data, (byte) 'x');
        File tmpDir = getTmpDir();

        try (final ChronicleQueue queue =
                     SingleChronicleQueueBuilder.builder(tmpDir, WireType.BINARY).blockSize(BLOCK_SIZE).build()) {

            for (int i = 0; i < 5; i++) {
                try (DocumentContext dc = queue.acquireAppender().writingDocument()) {
                    dc.wire().write("data").bytes(data);
                }

            }

        }

        try (final ChronicleQueue queue =
                     SingleChronicleQueueBuilder.builder(tmpDir, WireType.BINARY).blockSize(BLOCK_SIZE).build()) {

            for (int i = 0; i < 5; i++) {
                try (DocumentContext dc = queue.createTailer().readingDocument()) {
                    Assert.assertArrayEquals(data, dc.wire().read("data").bytes());
                }

            }

        }

    }

}
```

**Comment by RobAustin** [comment]
I will investigted if there is anything that can be done to prevent this from failing when the BLOCK_SIZE is smaller.

**Comment by gity** [comment]
@RobAustin this issue isn't just small block sizes. It happens on any block sizes with the certain number of bytes being written.

eg.
1048576 blockSize and 262036 bytes
67108864 blockSize and 16777216 bytes

The issue is in the original ticket, safeLength and safeCapacity are different between the queue and the mapped file that can cause writes past the mapped file. This could cause spurious SEGVs when certain messages are written.

I would have thought this would be something you would want to fix.

**Comment by RobAustin** [comment]
can you re-test with the latest snapshot as I feel this is now fixed

**Comment by hft-team-city** [comment]
Released in Chronicle-Queue-5.17.0 (https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.17.0), BOM-2.17.30 (https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-2.17.30)
